### PR TITLE
Implemented crater detection in Nvblox

### LINKED ISSUE #126: Add support for multiple esdf publishing
nvblox already has all esdf layers, we just need to make a ros wrapper to read them

### LINKED ISSUE #127: Add support for determining hole esdf layer
Obtain esdf once you have the below group height, then utilize this data to then determine where holes are.

### LINKED ISSUE #128: Add support for merging 2 costmaps into a single costmap for nav2
Once you have obstacle data from the below ground esdf determine and add in the obstacles to the costmap layer

### LINKED ISSUE #175: Figure out how to super impose 2 different esdf slices
The 2 esdf slices that we are using have different data sizes, and different origins. We need to normalize it so that the 2 matrices are aligned one on top of the other before merging them into a single costmap.

### LINKED ISSUE #176: Inverse the esdf point cloud (different from the nav2 map slice)
the esdf point cloud is the visual aspect of t costmap in Rviz. By inverting and republishing this while also republishing the regular costmap, we have a more visually appealing/usable outputhe

### PR DESCRIPTION
**Pull request by Isopod00**
This PR is ready for review now!

### CODE REVIEW

File: src/combine_esdf/package.xml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>combine_esdf</name>
+  <version>0.1.0</version>
+  <description>TODO: Package description</description>
+  <maintainer email="aaberg333@gmail.com">alex</maintainer>
+  <license>TODO: License declaration</license>
+
+  <test_depend>ament_copyright</test_depend>
```
Comment (jmblixt3) [inline src/combine_esdf/package.xml:10]: Add dependecies here

File: src/combine_esdf/setup.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        ('share/ament_index/resource_index/packages',
+            ['resource/' + package_name]),
+        ('share/' + package_name, ['package.xml']),
+    ],
+    install_requires=['setuptools'],
+    zip_safe=True,
+    maintainer='alex',
+    maintainer_email='aaberg333@gmail.com',
+    description='Combines 2 esdf maps into a single esdf map',
+    license='TODO: License declaration',
```
Comment (jmblixt3) [inline src/combine_esdf/setup.py]: Set licence

File: src/combine_esdf/combine_esdf/combine_esdf.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        if (self.below_ground_costmap_one is None):
+            self.costmap_publisher.publish(msg)
+            return
+        self.above_ground_costmap = np.array(msg.data).reshape(msg.height, msg.width)
+        # print(self.above_ground_costmap.shape, msg.origin, msg.resolution)
+        
+        
+        # self.above_ground_costmap = np.array(msg.data).reshape(msg.height, msg.width)
+        # # Set any null values in the costmap to an inaccessible area (before inverting)
+        condition = self.below_ground_costmap_one == 0
```
Comment (jmblixt3) [inline src/combine_esdf/combine_esdf/combine_esdf.py]: Don't use double equal sign

File: src/combine_esdf/combine_esdf/combine_esdf.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        origin = [msg.origin.x, msg.origin.y]
+
+        # self.above_ground_costmap[]
+        if (self.above_ground_costmap.shape[0] > self.below_ground_costmap_one.shape[0] and self.above_ground_costmap.shape[1] > self.below_ground_costmap_one.shape[1]):
+            try:
+                offset_x1 = abs(int((origin[0] - self.below_origin[0]) / msg.resolution))
+                offset_y1 = abs(int((origin[1] - self.below_origin[1]) / msg.resolution))
+                offset_x2 = self.above_ground_costmap.shape[0] - ((self.above_ground_costmap.shape[0] - self.below_ground_costmap_one.shape[0]) - offset_x1)
+                offset_y2 = self.above_ground_costmap.shape[1] - ((self.above_ground_costmap.shape[1] - self.below_ground_costmap_one.shape[1]) - offset_y1)
+                print(offset_x1, offset_x2, self.above_ground_costmap.shape, self.below_ground_costmap_one.shape)
```
Comment (jmblixt3) [inline src/combine_esdf/combine_esdf/combine_esdf.py]: Change print statements to roslog statements

**Review by jmblixt3 — COMMENTED** [review]
Some findings

File: src/combine_esdf/setup.py (hunk 2 above)
Comment (Isopod00) [reply src/combine_esdf/setup.py]: MIT License is my go-to @BergerKing333 :)

File: src/combine_esdf/package.xml (hunk 1 above)
Comment (Isopod00) [reply src/combine_esdf/package.xml:10]: Set License here as well

File: numpy_matrix_image.png
Comment (Isopod00) [inline numpy_matrix_image.png:1]: please delete this image

File: numpy_matrix_image2.png
Comment (Isopod00) [inline numpy_matrix_image2.png:1]: please delete this image

File: src/combine_esdf/package.xml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  <version>0.1.0</version>
+  <description>Combines 2 Nvblox ESDF slices into a single ESDF slice</description>
+  <maintainer email="aaberg333@gmail.com">Alex Berg</maintainer>
+  <license>MIT License</license>
+
+  <test_depend>ament_copyright</test_depend>
+  <test_depend>ament_flake8</test_depend>
+  <test_depend>ament_pep257</test_depend>
+  <test_depend>python3-pytest</test_depend>
+  <test_depend>numpy</test_depend>
```
Comment (Isopod00) [inline src/combine_esdf/package.xml]: for numpy and scipy change them from "test_depend" to either "depend" or "exec_depend"

File: .gitignore [hunk 6]
```diff
@@ -55,4 +55,5 @@ qtcreator-*
 # Catkin custom files
 CATKIN_IGNORE
 
-.fake
\ No newline at end of file
+.fake
```
Comment (jmblixt3) [inline .gitignore]: remove ssh

File: src/combine_esdf/combine_esdf/combine_esdf.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+            offset_x1 = abs(int((origin[1] - self.below_origin[1]) / msg.resolution))
+            offset_y1 = abs(int((origin[0] - self.below_origin[0]) / msg.resolution))
+            offset_x2 = ((self.above_ground_costmap.shape[0] - self.below_ground_costmap.shape[0]) - offset_x1)
+            offset_y2 = ((self.above_ground_costmap.shape[1] - self.below_ground_costmap.shape[1]) - offset_y1)
+            
+            # To invert or not to invert, that is the question. 
+            # Comment this next chunk out if the lower esdf is only showing the hole as red, not the terrain.
+            # this actually doesn't work yet, so leave it commented.
+            # self.below_ground_costmap[self.below_ground_costmap == 1000] = 0
+            # self.below_ground_costmap = np.full(self.below_ground_costmap.shape, 2) - self.below_ground_costmap
```
Comment (jmblixt3) [inline src/combine_esdf/combine_esdf/combine_esdf.py]: don't put a comment make an if statement with a config

File: src/combine_esdf/combine_esdf/combine_esdf.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+            pass
+
+    # Below ground costmap callback; sets the variables for below ground costmap and origin.
+    # Easily replicable for more below ground costmaps. Stack the costmaps into a 3D array
+    # Probably useful to move the inversion logic to this function if this is the case.
+    def below_ground_callback(self, msg):
+        if (len(msg.data) == 0):
+            return   
+        self.below_ground_costmap = np.array(msg.data).reshape(msg.height, msg.width)
+        self.below_origin = [msg.origin.x, msg.origin.y]
```
Comment (jmblixt3) [inline src/combine_esdf/combine_esdf/combine_esdf.py]: can you not use a tuple here

File: src/combine_esdf/combine_esdf/combine_esdf.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+            1)
+        
+        self.below_ground_subscriber = self.create_subscription(
+            DistanceMapSlice,
+            '/nvblox_node/static_map_slice2',
+            self.below_ground_callback,
+            1)
+        
+
+        # TODO: I'd like to put these in an if loop based on a config file parameter
```
Comment (jmblixt3) [inline src/combine_esdf/combine_esdf/combine_esdf.py]: Are you going to do this

**Review by jmblixt3 — DISMISSED** [review]
Some things to fix/consider?

**Comment by Isopod00** [comment]
Before merging this, please modify the nav2_launch.py launch file to launch your combine_esdf node

File: src/isaac_ros/isaac_ros_launch/launch/nvblox.launch.py [hunk 10]
```diff
@@ -105,4 +105,10 @@ def generate_launch_description():
         ]
     )
 
-    return LaunchDescription([nvblox_container, group_action])
+    combine_esdf = Node(
```
Comment (Isopod00) [inline src/isaac_ros/isaac_ros_launch/launch/nvblox.launch.py:108]: test this launch file change to make sure it works

**Review by BergerKing333 — APPROVED** [review]
anthony said this is good

**Pull request merged**